crates/egui_demo_lib/src/demo/pan_zoom.rs

```diff
@@ -0,0 +1,102 @@
+use egui::emath::TSTransform;
+
+#[derive(Clone, Default, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct PanZoom {
+    transform: TSTransform,
+}
+
+impl Eq for PanZoom {}
+
+impl super::Demo for PanZoom {
+    fn name(&self) -> &'static str {
+        "🗖 Pan Zoom"
+    }
+
+    fn show(&mut self, ctx: &egui::Context, open: &mut bool) {
+        use super::View as _;
+        let window = egui::Window::new("Pan Zoom")
+            .default_width(200.0)
+            .default_height(200.0)
+            .vscroll(false)
+            .open(open);
+        window.show(ctx, |ui| self.ui(ui));
+    }
+}
+
+impl super::View for PanZoom {
+    fn ui(&mut self, ui: &mut egui::Ui) {
+        let (id, rect) = ui.allocate_space(ui.available_size());
+        let response = ui.interact(rect, id, egui::Sense::click_and_drag());
+        // Uncomment to allow dragging the background as well.
+        // self.transform.translation += response.drag_delta() / self.transform.scaling;
+
+        // Plot-like reset
+        if response.double_clicked() {
+            self.transform = TSTransform::default();
+        }
+
+        if let Some(pointer) = ui.ctx().input(|i| i.pointer.hover_pos()) {
+            // Note: doesn't catch zooming / panning if a button in this PanZoom container is hovered.
+            if response.hovered() {
+                let original_zoom = self.transform.scaling;
+                let new_zoom = original_zoom * ui.ctx().input(|i| i.zoom_delta());
+
+                let layer_pos = self.transform.inverse() * pointer;
+                let new_transform = TSTransform::new(self.transform.translation, new_zoom);
+                let new_pos = new_transform * layer_pos;
+
+                // Keep mouse centered.
+                let pan_delta = pointer - new_pos;
+
+                // Handle scrolling.
+                let pan_delta =
+                    pan_delta + ui.ctx().input(|i| i.smooth_scroll_delta) / original_zoom;
+
+                self.transform.scaling = new_zoom;
+                self.transform.translation += pan_delta;
```

Comment:
A simpler way of expressing this would be something like:

```
let pointer_in_layer = self.transform.inverse() * pointer;
let zoom_delta = ui.ctx().input(|i| i.zoom_delta());
let pan_delta = ui.ctx().input(|i| i.smooth_scroll_delta);

// Zoom in on pointer:
self.transform =
    self.transform *
    TSTransform::translate(pointer_in_layer) *
    TSTransform::scale(zoom_delta) *
    TSTransform::translate(-pointer_in_layer);

// Pan:
self.transform = TSTransform::translate(pan_delta) * self.transform;
```

There might be a negation sign wrong there, but the gist of it should be correct.

Reply:
Ahh, that's very clever and much more readable!